### PR TITLE
Enable bloom filters and add prefix extractor.

### PR DESCRIPTION
**Pull request by petermattis**
The `<user-key>` portion of MVCC keys is now extracted and added to the
per table bloom filter which allows skipping of tables which do not
contain the `<user-key>` during "point" lookups (i.e. MVCCGet).

Enabled rocksdb statistics collection. We should probably bounce some of
the statistics up to Go-land. Log the bloom prefix filter utility when
closing a DB which shows if the bloom filter is being utilized or not.

Changed the way that mvcc data is generated for benchmarks. The keys are
now written in random order and the final compaction step is
disabled. The result is a series of sstable files instead of just 1 and
the keys randomly distributed through the sstables.

```
name                      old time/op    new time/op     delta
MVCCGet1Version8Bytes        104µs ± 4%       35µs ± 1%   -66.34%         (p=0.000 n=10+9)
MVCCGet10Versions8Bytes     89.6µs ± 2%     47.5µs ± 4%   -46.95%         (p=0.000 n=10+9)
MVCCGet100Versions8Bytes    82.4µs ± 7%     78.3µs ± 1%    -4.98%        (p=0.000 n=10+10)
```

The above benchmark numbers are comparing with prefix iteration disabled
vs enabled. The 100 version test shows little benefit because the
versions get spread throughout the sstables removing much of the utility
of the bloom filtering.

Removed the non-8-byte versions of the MVCCGet tests as they were simply
showing the benefit of reading more data at once. Removed parallelism
from the MVCCGet benchmarks as it was obfuscating the performance of a
single operation. Disabled the rocksdb cache in order to highlight the
performance when "reading from disk" (usually the OS buffer cache).

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3284)

### CODE REVIEW

**Comment by tamird** [comment]
> The portion of MVCC keys is now extracted and added to the
per table bloom filter which allows skipping of tables which do not
contain the during "point" lookups (i.e. MVCCGet).

Which portion? Tables which contain the what?

---

Reviewed 10 of 11 files at r1.
Review status: 10 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/engine.go, line 135 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YqqBh1uDchoLB_Aki:-K4YqqBh1uDchoLB_Akj:1032658164) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/engine.go#L135)):*
the same MVCC user-key prefix as what? Each other? the first call to either? the first call to seek?

---

*[storage/engine/rocksdb/db.h, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YrAN-ueEgJzibfXbH:-K4YrAN-ueEgJzibfXbI:1546512336) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb/db.h#L119)):*
same question: the same MVCC user-key prefix as what? Each other? the first call to either? the first call to seek?

---

*[storage/engine/rocksdb_test.go, line 166 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YreYtvlx2zrbHz7w_:-K4YreYtvlx2zrbHz7wa:1772900795) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L166)):*
super nit: why not `for j := 0; j < keyVersions; j++`? makes no difference, just that that form is much more common (and starting at 1 makes me look for a reason, which there doesn't appear to be).

---

*[storage/engine/rocksdb_test.go, line 172 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Ys57IEa-lZXwG5GGU:-K4Ys57IEa-lZXwG5GGV:-115531279) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L172)):*
I'm not sure if it matters, but [fisher-yates](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle) is more like:
```go
n := len(order)
for i := range order[:n-2] {
    j := rand.Intn(n-i) + i
    a[i], a[j] = a[j], a[i]
}
```

---

*[storage/engine/rocksdb_test.go, line 179 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Yu3wjxoWtt7C6mv4E:-K4Yu3wjxoWtt7C6mv4F:-1049500665) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L179)):*
can you add a comment explaining why it's important that many batches are used?

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4YuHNB05wk9FCVVkc4:312083747) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
what's going on here with `j`?

---

*[storage/engine/rocksdb_test.go, line 357 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Yubr-DIS3StBkXCGF:-K4Yubr-DIS3StBkXCGG:-380139138) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L357)):*
can you break out `512` and `1000000` into name constants?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-:-K4YqbeebQboCRJrlQQN:876705165)*

**Comment by petermattis** [comment]
Hmm, `<user-key>` seems to be stripped from the text in some contexts. Try reading the original commit message. 

---

Review status: 10 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/engine.go, line 135 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YqqBh1uDchoLB_Aki:-K4YyDuVaf19Te484gVs:-1351396131) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/engine.go#L135)):*
An MVCC key is composed of `user-key/timestamp`. When `prefix` is true `Seek` will use the `user-key` prefix  to restrict which sstables are searched, but iteration (using `Next`) over keys without the same `user-key`will not work correctly. I've updated the comment.

---

*[storage/engine/rocksdb/db.h, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YrAN-ueEgJzibfXbH:-K4Yz29__ucwsBHrVJ2Y:1582050969) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb/db.h#L119)):*
Updated comment.

---

*[storage/engine/rocksdb_test.go, line 166 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YreYtvlx2zrbHz7w_:-K4Yz4as_CR2mdcKEQjW:1901739485) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L166)):*
The 0 timestamp is special.

---

*[storage/engine/rocksdb_test.go, line 172 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Ys57IEa-lZXwG5GGU:-K4YzOtUpHmL9Z4_um3G:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L172)):*
Done.

---

*[storage/engine/rocksdb_test.go, line 179 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Yu3wjxoWtt7C6mv4E:-K4YzrmzxsndQlKdelXr:-149651698) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L179)):*
Done. If we used a single batch rocksdb would create a single sstable. We want many sstables.

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4Yzgtbwd1w9mLnE0h8:-820398735) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
The 3rd portion of a for-loop is a single statement. `i++` is a statement. I suppose I could do `i, j = i+1, j+1`

---

*[storage/engine/rocksdb_test.go, line 357 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Yubr-DIS3StBkXCGF:-K4Z-O6N6CUz63p9ovzo:403151359) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L357)):*
Done. It's a target size for the data.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-:-K4YxYKgNltT1WeSXzEl:-1952901722)*

**Comment by tamird** [comment]
Probably has special meaning in markdown.

---

Reviewed 3 of 3 files at r2.
Review status: 10 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/engine/engine.go, line 137 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Z2GtzzDWjLWa2QBn5:-K4Z2Gu-QxLk-gNZ1M9Y:121796480) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ed155b0406a4191185bfd931e2504a88a3de025/storage/engine/engine.go#L137)):*
> will not work correctly

What will happen? undefined?

---

*[storage/engine/rocksdb/db.h, line 121 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Z2NKYDUgu6uBupT1Y:-K4Z2NKYDUgu6uBupT1Z:2133697532) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ed155b0406a4191185bfd931e2504a88a3de025/storage/engine/rocksdb/db.h#L121)):*
same question re what will happen

---

*[storage/engine/rocksdb_test.go, line 166 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YreYtvlx2zrbHz7w_:-K4Z2SsQMJ5uWqR_xJml:-177959318) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L166)):*
I still don't follow, `j` is not actually being used for anything other than the loop counter.

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4Z2pFvkxqPyMrsgHJI:-1675616442) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
That's cleaner, but I meant, why is `j` capped at `100`? What's `j` being used to track here?

---

*[storage/engine/rocksdb_test.go, line 357 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Yubr-DIS3StBkXCGF:-K4Z3-U_u6ROozXck-4A:836696371) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L357)):*
Much better, thanks.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-:-K4Z-npyGBWh9IzhofSK:1523315884)*

**Comment by petermattis** [comment]
Review status: 8 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/engine/engine.go, line 137 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4Z2GtzzDWjLWa2QBn5:-K4Z3IYENg90rMVNEEPZ:-1871472533) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ed155b0406a4191185bfd931e2504a88a3de025/storage/engine/engine.go#L137)):*
Keys might be skipped.

---

*[storage/engine/rocksdb_test.go, line 166 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YreYtvlx2zrbHz7w_:-K4Z3YMiW-Cpu61Of36M:-407972374) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L166)):*
Ah, good point. Previous version of this code was appending `j` to a slice, but then shuffling would mess up the order of insertion of the versions. Switched to the more natural for-loop.

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4Z3ksRuNyDwDLJE1XR:-1543448424) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
`100` is the batch size.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284)*

**Comment by tamird** [comment]
Reviewed 2 of 3 files at r2, 3 of 3 files at r3.
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4Z62RaYzJaEQCDpwuo:2063556554) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
I find the current code very hard to read. Is the below equivalent?

```go
batch := rocksdb.NewBatch()
for i, idx := range order {
   if i % 100 == 0 {
     batch.Commit()
     batch = rocksdb.NewBatch()
   }
   ....
}
batch.Commit()
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284)*

File: storage/engine/rocksdb_test.go [hunk 1]
```diff
@@ -135,15 +135,15 @@ func TestRocksDBCompaction(t *testing.T) {
...
+func setupMVCCData(numVersions, numKeys, valueBytes int, b *testing.B) (*RocksDB, *stop.Stopper) {
+	loc := fmt.Sprintf("mvcc_data_%d_%d_%d", numVersions, numKeys, valueBytes)
 
 	exists := true
 	if _, err := os.Stat(loc); os.IsNotExist(err) {
 		exists = false
 	}
 
-	const cacheSize = 8 << 30 // 8 GB
+	const cacheSize = 0
```
Comment (tschottdorf) [inline storage/engine/rocksdb_test.go]: why keep this?

**Comment by petermattis** [comment]
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4Z8jHTIa3yEAW3LNta:-632968060) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
Almost. That's committing on the first iteration unnecessarily. I suppose I can restructure this. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284)*

File: storage/engine/rocksdb_test.go (hunk 1 above)
Comment (petermattis) [inline storage/engine/rocksdb_test.go]: No good reason. It does document the parameter, though.

**Comment by tamird** [comment]
LGTM but did not review C++ changes.

---

Reviewed 1 of 1 files at r4.
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/engine/rocksdb_test.go, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-K4YuHNA89C_zBuJoBSi:-K4ZDSS9GE_4_K06_Ft5:-1573950660) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d6a1efb6bd0c42f98fa4670e21aac5bfc90d08/storage/engine/rocksdb_test.go#L180)):*
Thanks, that's much clearer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-:-K4ZDVaNNZ2pmjxFY_M_:1823416636)*

**Comment by petermattis** [comment]
Anyone want to take a gander at the C++? The changes are fairly straightforward, with `DBPrefixExtractor` being the interesting bit.

---

Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3284#-:-K4ZDnfd-e3-rFEQMlRF:1514896854)*

File: storage/engine/rocksdb/db.cc [hunk 2]
```diff
@@ -1089,6 +1130,8 @@ DBStatus DBOpen(DBEngine **db, DBSlice dir, DBOptions db_opts) {
     table_options.block_cache = rocksdb::NewLRUCache(
         block_cache_size, num_cache_shard_bits);
   }
+  table_options.filter_policy.reset(rocksdb::NewBloomFilterPolicy(10));
+  table_options.format_version = 2;
```
Comment (bdarnell) [inline storage/engine/rocksdb/db.cc]: Does changing the format do anything besides (presumably) enabling the prefix-based bloom filter? Is it worth benchmarking this change separately?

Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: No, changing the format did nothing. It isn't even necessary for prefix-based bloom filters. This was a drive-by changed based on the following comment in the rocksdb source:

```
  // We currently have three versions:
  // 0 -- This version is currently written out by all RocksDB's versions by
  // default.  Can be read by really old RocksDB's. Doesn't support changing
  // checksum (default is CRC32).
  // 1 -- Can be read by RocksDB's versions since 3.0. Supports non-default
  // checksum, like xxHash. It is written by RocksDB when
  // BlockBasedTableOptions::checksum is something other than kCRC32c. (version
  // 0 is silently upconverted)
  // 2 -- Can be read by RocksDB's versions since 3.10. Changes the way we
  // encode compressed blocks with LZ4, BZip2 and Zlib compression. If you
  // don't plan to run RocksDB before version 3.10, you should probably use
  // this.
  // This option only affects newly written tables. When reading exising tables,
  // the information about version is read from the footer.
```

I did benchmark this change in isolation and it did not result in any performance change.

File: storage/engine/rocksdb/db.cc [hunk 3]
```diff
@@ -324,6 +332,39 @@ class DBCompactionFilterFactory : public rocksdb::CompactionFilterFactory {
...
+
+  // MVCC keys are encoded as <user-key>/<timestamp>. Extract the <user-key>
+  // prefix which will allow for more efficient iteration over the keys
+  // matching a particular <user-key>. Specifically, the <user-key> will be
+  // added to the per table bloom filters and will this be able to skip tables
+  // which do not contain the <user-key>.
+  virtual rocksdb::Slice Transform(const rocksdb::Slice& src) const {
+    const char kTerm[] = "\x00\x01";
+    const int kTermSize = sizeof(kTerm) - 1;
+    const char* p = reinterpret_cast<const char*>(memmem(src.data(), src.size(), kTerm, kTermSize));
```
Comment (bdarnell) [inline storage/engine/rocksdb/db.cc]: Why do we need to search for `\x00\x01` instead of just removing the last 12 bytes?

**Comment by bdarnell** [comment]
LGTM (including the c++)

File: storage/engine/rocksdb/db.cc (hunk 3 above)
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Because keys for `MVCCMetadata` values do not contain the 12-byte timestamp suffix.

Comment (bdarnell) [inline storage/engine/rocksdb/db.cc]: Ah, OK. I was thinking they had a suffix of 12 zeros.
